### DISCUSSION TITLE
ESP32-WROOM-32: wrong firmware or wifi guru meditation?

### DISCUSSION THREAD

**Discussion by sysopch** [micropython/micropython#11031] [ESP32]
Hi all,
I have some previous experience with ESP32 and python, but I am pretty new to micropython. So I'd like to double check that I am running the correct firmware first.
I have a cheap board integrating ESP32-WROOM-32 (there is a XXSR69 code on the module, if this helps), which as far as I understand integrates chip ESP32-D0WDQ6.
As this is a dual core chip I first tried downloading the ESP32 port with this command
`esptool.py --chip esp32 --port /dev/ttyUSB0 write_flash -z 0x1000 ~/shared/esp32-20220618-v1.19.1.bin`
This results in a bootloop

> rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12344
ho 0 tail 12 room 4
load:0x40080400,len:4124
entry 0x40080680
Guru Meditation Error: Core  0 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x40096ec4: fd56d132 002850b1 bf42a850
Core  0 register dump:
PC      : 0x40096ecb  PS      : 0x00060b30  A0      : 0x800970b8  A1      : 0x3ffe3ba0  
A2      : 0x3ffb7880  A3      : 0x3ffb7880  A4      : 0x00000000  A5      : 0x00000000  
A6      : 0x3ffc3bf0  A7      : 0x00000020  A8      : 0x80096e08  A9      : 0x3ffe3ba0  
A10     : 0x00000001  A11     : 0x0002e002  A12     : 0x3ffb78cc  A13     : 0x04b45700  
A14     : 0x00000004  A15     : 0x0000000a  SAR     : 0x0000001d  EXCCAUSE: 0x00000000  
EXCVADDR: 0x00000000  LBEG    : 0x400827f5  LEND    : 0x400827fd  LCOUNT  : 0x00000000  
Backtrace:0x40096ec8:0x3ffe3ba0 0x400970b5:0x3ffe3be0 0x400d3071:0x3ffe3c00 0x40081a39:0x3ffe3c20 0x40081c90:0x3ffe3c70 0x40078f00:0x3ffe3c90 |<-CORRUPTED

The 'CORRUPTED' message suggest some flashing issue, however `esptool verify_flash` returns
> Verifying 0x17d190 (1560976) bytes @ 0x00001000 in flash against /home/aldo/shared/esp32-20220618-v1.19.1.bin...
-- verify OK (digest matched)

So I assumed this is not the correct FW and went for the UNICORE version.

Loading `esptool.py --chip esp32 --port /dev/ttyUSB0 write_flash -z 0x1000 ~/shared/esp32-unicore-20230313-unstable-v1.19.1-966-g05bb26010.bin` seems to work, I can connect with a terminal and I was even able to connect jupyter notebook and upload files with ampy.

However I get a guru meditation if I try to configure wifi with

`import network`
`ap = network.WLAN(network.AP_IF)`

>Guru Meditation Error: Core  0 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x40096ec4: fd009525 304c4402 9f6502ad
Core  0 register dump:
PC      : 0x40096ec9  PS      : 0x00060b30  A0      : 0x8014b698  A1      : 0x3ffcc660  
A2      : 0x3ffbc0d0  A3      : 0x00004289  A4      : 0x3ffbc11c  A5      : 0x3ffbc0d0  
A6      : 0x00000001  A7      : 0x3ffbc0f4  A8      : 0x80081ac4  A9      : 0x3ffcc640  
A10     : 0x3ff000dc  A11     : 0x00000001  A12     : 0x3ffbdde8  A13     : 0x00000000  
A14     : 0x00000001  A15     : 0x3ffbc0d0  SAR     : 0x0000001f  EXCCAUSE: 0x00000000  
EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0xffffffff  
>
>Backtrace:0x40096ec6:0x3ffcc660 0x4014b695:0x3ffcc680 0x401b4cdd:0x3ffcc6a0 0x401af3e7:0x3ffcc6e0 0x40192f4a:0x3ffcc700 0x4019305e:0x3ffcc720 0x4014b04e:0x3ffcc740 0x400fae71:0x3ffcc760 0x400dd007:0x3ffcc840 0x400e3d6a:0x3ffcc870 0x400e3e99:0x3ffcc890 0x400842d5:0x3ffcc8b0 0x400dcf9a:0x3ffcc950 0x400e3d6a:0x3ffcc9a0 0x400e3d92:0x3ffcc9c0 0x400f19b2:0x3ffcc9e0 0x400f1cec:0x3ffcca70 0x400d5cfc:0x3ffccab0
>
>
>ELF file SHA256: c860aa869524a60e
>
>Rebooting...

What am I doing wrong?

NOTE I found some notes about adding a cap between EN and GND, but as far as I understand (also from previous experience) this should be only related to instability in switching to the bootloader. I am not having download errors, just runtime, so I assume this is not the problem.

**Comment by davefes** [comment]
Curious where did you get:
`/esp32-unicore-20230313-unstable-v1.19.1-966-g05bb26010.bin`
from?

Does this tiny board have the transistor pair and a boot button?  Are you using rshell on a Linux platform or an IDE?

**Reply by sysopch** [reply]
I downloaded the latest from this page https://micropython.org/download/esp32-unicore/

Yes, the board has RESET and BOOT buttons, I assume also the transistors for the boot mode switch.
The board is [this](https://www.jzk-jzk.com/products/esp-32s-esp32-development-board-2-4ghz-wifi-bluetooth-dual-mode-antenna-module-with-ultra-low-power) 
![image](https://user-images.githubusercontent.com/49515572/225170435-6d81c9c0-903c-46fa-ba94-48e9be4315d7.png)

I am on linux with esptool in a virtual env

**Reply by davefes** [reply]
Hmm, I thought unicore would be for the new RISC-based variants.

I am using the D0WDQ6 but rev 1 chips.  Which rev are yours?

`esptool.py  --chip esp32 chip_id`

**Reply by sysopch** [reply]
`Warning: ESP32 has no Chip ID. Reading MAC instead.
MAC: e8:68:e7:14:a7:24`

If I found the correct doc, the one of the first that Express has made for wroom and is Now not recommended for New Designs, so I guess That is rev 0

About unicore I thought that would be the version for single core variants, the chips which start with S, but I may be mistaken

**Reply by sysopch** [reply]
I take it back, I have rev.1. esptool says that when connecting (didn't notice before):

> Serial port /dev/ttyUSB0
Connecting....
Chip is ESP32-D0WD (revision v1.0)

even if it cites a different chip... I am a little puzzled here

**Reply by robert-hh** [reply]
So you can try the basic dual core firmware. Using the wrong firmware does not damage anything, since the boot loader is in ROM.

**Reply by sysopch** [reply]
we replied at the same time ;)
see my answer below your suggestion, it does not work

**Comment by robert-hh** [comment]
The picture is good. I see two transistors, and there is output after booting, so the boot logic is there and works. But the ESP32-WROOM is AFAIK the standard two core ESP32. So the basic firmware, should work,  e.g. https://micropython.org/resources/firmware/esp32-20230310-unstable-v1.19.1-962-g4376c969f.bin

Edit: @davefes  There are single core variants of the ESP32, but rarely used. And the RISC version is the ESP32-C3.

**Reply by sysopch** [reply]
No luck, still bootloop

> Rebooting...
ets Jun  8 2016 00:22:57
>
>rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4656
load:0x40078000,len:13284
ho 0 tail 12 room 4
load:0x40080400,len:3712
entry 0x4008064c
Guru Meditation Error: Core  0 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x40096ed8: 80002136 d9e43003 20002010
Core  0 register dump:
PC      : 0x40096ede  PS      : 0x00060b33  A0      : 0x000000b8  A1      : 0x3ffe3b20  
A2      : 0x00000004  A3      : 0x3ff001bc  A4      : 0x00000002  A5      : 0xffffffff  
A6      : 0x00000001  A7      : 0x00000000  A8      : 0x3ff00104  A9      : 0x3ffbeab0  
A10     : 0x3ffaff40  A11     : 0xfffeffff  A12     : 0x00040000  A13     : 0x00000000  
A14     : 0x00000804  A15     : 0x00000000  SAR     : 0x0000001e  EXCCAUSE: 0x00000000  
EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000000  
>
>Backtrace:0x40096edb:0x3ffe3b20 0x000000b5:0x3ffe3b30 |<-CORRUPTED
>
>
>ELF file SHA256: 620f574b23fec714
>
>Rebooting...

**Reply by davefes** [reply]
Was not aware of single core variants of the ESP32.

> D0WD is a smaller version of the D0WDQ6 (Wikipedia)

**Reply by robert-hh** [reply]
I would try now different flash modes. That was required for ESP8266 and I never heard that it was required for an ESP32, but ....

**Comment by robert-hh** [comment]
I hate the layout of discussions. Answers pop up at different places.

**Comment by sysopch** [comment]
Let's try to go in sequence. (I mean with the order of the replies)

flash modes: I am not sure if I understand you correctly. As I tried verify and it did work, I assumed flashing is ok, is this correct?

I guess we can always go back right?
What values do you suggest to try?

**Comment by davefes** [comment]
> However I get a guru meditation if I try to configure wifi with

```py
import network
ap = network.WLAN(network.AP_IF)
```
All my boards I put a 1000uF cap on the 3V3.  If the WiFi is really trying to boot-up at this point you can get "dips" on the 3V3 line.
How are you powering the board?

**Comment by sysopch** [comment]
> All my boards I put a 1000uF cap on the 3V3. If the WiFi is really trying to boot-up at this point you can get "dips" on the 3V3 line.

ok, I'll try that later at work

**Reply by robert-hh** [reply]
There are four flash modes. You tried already DIO, so the next would be DOUT, then QIO and QOUT. The the other options are --flash_freq. AFAIK, the default is 80Mhz. You can try 40Mhz.

**Comment by robert-hh** [comment]
> All my boards I put a 1000uF cap on the 3V3. 

I have one cheap ESP32 board purchased recently to check the error with a 5x5 mm ESP32. I got the wrong one, but that required a capacitor between EN and GND, when starting WiFi. But it booted fine.
In fact I added a 1µ capacitor between EN and 3V3, just because I could simply solder a SMD cap between two adjacent pins. But your board's pin arrangement is different.

**Comment by sysopch** [comment]
Summary of my attempts:
- I put a 470uF cap between 3.3 and GND. I thought I have seen wifi init working once (` network.WLAN(network.AP_IF)` without crash. I am not able to repro. It always crashes now with unipro, bootloop with generic esp32
- I tried with dio and dout flash modes, no luck
- I tried rebuilding the GENERIC port myself, still bootloop
- I tried some examples from esp-idf wifi (scan, getting_started). It appears it is able to initialize wifi but does not join to my AP (I triple-checked SSID and password, but I may still be making some silly mistake)

If I will find the time tomorrow I will try to monitor the power line with an oscilloscope and see if I detect any glitches.
@robert-hh  you mentioned that you put a cap between EN and GND, even if booting was fine, rigtht? I may try that too.

Let me paste the `idf.py monitor` output here, just in case it contains any useful hint

> -- idf_monitor on /dev/ttyUSB1 115200 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
ets Jun  8 2016 00:22:57
>
>rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:6676
load:0x40078000,len:14944
ho 0 tail 12 room 4
load:0x40080400,len:3816
0x40080400: _init at ??:?
>
>entry 0x40080698
I (28) boot: ESP-IDF v4.4.4 2nd stage bootloader
I (28) boot: compile time 00:34:24
I (29) boot: chip revision: v1.0
I (31) boot_comm: chip revision: 1, min. bootloader chip revision: 0
I (39) boot.esp32: SPI Speed      : 40MHz
I (43) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 2MB
I (52) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (61) boot: ## Label            Usage          Type ST Offset   Length
I (69) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (76) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (83) boot:  2 factory          factory app      00 00 00010000 00100000
I (91) boot: End of partition table
I (95) boot_comm: chip revision: 1, min. application chip revision: 0
I (102) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=12e20h ( 77344) map
I (139) esp_image: segment 1: paddr=00022e48 vaddr=3ffb0000 size=038a8h ( 14504) load
I (145) esp_image: segment 2: paddr=000266f8 vaddr=40080000 size=09920h ( 39200) load
I (161) esp_image: segment 3: paddr=00030020 vaddr=400d0020 size=6f420h (455712) map
I (326) esp_image: segment 4: paddr=0009f448 vaddr=40089920 size=0b314h ( 45844) load
I (355) boot: Loaded app from partition at offset 0x10000
I (356) boot: Disabling RNG early entropy source...
I (367) cpu_start: Pro cpu up.
I (367) cpu_start: Starting app cpu, entry point is 0x400811c4
0x400811c4: call_start_cpu1 at /home/.../esp-idf/components/esp_system/port/cpu_start.c:148
>
>I (0) cpu_start: App cpu up.
I (381) cpu_start: Pro cpu start user code
I (381) cpu_start: cpu freq: 160000000
I (381) cpu_start: Application information:
I (386) cpu_start: Project name:     wifi_station
I (391) cpu_start: App version:      v4.4.4
I (396) cpu_start: Compile time:     Mar 16 2023 00:41:16
I (402) cpu_start: ELF file SHA256:  b83ab0ff642a0c8f...
I (408) cpu_start: ESP-IDF:          v4.4.4
I (413) heap_init: Initializing. RAM available for dynamic allocation:
I (420) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (426) heap_init: At 3FFB7280 len 00028D80 (163 KiB): DRAM
I (433) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (439) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (445) heap_init: At 40094C34 len 0000B3CC (44 KiB): IRAM
I (453) spi_flash: detected chip: generic
I (456) spi_flash: flash io: dio
W (460) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (474) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (568) wifi station: ESP_WIFI_MODE_STA
I (588) wifi:wifi driver task: 3ffbf924, prio:23, stack:6656, core=0
I (588) system_api: Base MAC address is not set
I (588) system_api: read default base MAC address from EFUSE
I (618) wifi:wifi firmware version: 6567a16
I (618) wifi:wifi certification version: v7.0
I (618) wifi:config NVS flash: enabled
I (618) wifi:config nano formating: disabled
I (618) wifi:Init data frame dynamic rx buffer num: 32
I (618) wifi:Init management frame dynamic rx buffer num: 32
I (628) wifi:Init management short buffer num: 32
I (628) wifi:Init dynamic tx buffer num: 32
I (638) wifi:Init static rx buffer size: 1600
I (638) wifi:Init static rx buffer num: 10
I (648) wifi:Init dynamic rx buffer num: 32
I (648) wifi_init: rx ba win: 6
I (648) wifi_init: tcpip mbox: 32
I (658) wifi_init: udp mbox: 6
I (658) wifi_init: tcp mbox: 6
I (668) wifi_init: tcp tx win: 5744
I (668) wifi_init: tcp rx win: 5744
I (668) wifi_init: tcp mss: 1440
I (678) wifi_init: WiFi IRAM OP enabled
I (678) wifi_init: WiFi RX IRAM OP enabled
I (688) phy_init: phy_version 4670,719f9f6,Feb 18 2021,17:07:07
I (798) wifi:mode : sta (e8:68:e7:14:a7:24)
I (798) wifi:enable tsf
I (798) wifi station: wifi_init_sta finished.
I (928) wifi:state: init -> auth (b0):<255,255>, sta:<2,0>, prof:1
I (938) wifi:state: auth -> assoc (0)
I (938) wifi:state: assoc -> run (10)
I (7038) wifi:state: run -> init (fc0)
`I (7038) wifi:new:<2,0>, old:<2,0>, ap:<255,255>, sta:<2,0>, prof:1`
I (7038) wifi station: retry to connect to the AP
I (7038) wifi station: connect to the AP fail
I (9448) wifi station: retry to connect to the AP
I (9448) wifi station: connect to the AP fail
`I (9458) wifi:new:<2,0>, old:<2,0>, ap:<255,255>, sta:<2,0>, prof:1`
I (9458) wifi:state: init -> auth (b0)
I (9458) wifi:state: auth -> assoc (0)
I (9468) wifi:state: assoc -> run (10)
I (15568) wifi:state: run -> init (fc0)
I (15568) `wifi:new:<2,0>, old:<2,0>, ap:<255,255>, sta:<2,0>, prof:1`
I (15568) wifi station: retry to connect to the AP
I (15568) wifi station: connect to the AP fail
I (17978) wifi station: retry to connect to the AP
I (17978) wifi station: connect to the AP fail
I (17988)  `wifi:new:<2,0>, old:<2,0>, ap:<255,255>, sta:<2,0>, prof:1`
I (17988) wifi:state: init -> auth (b0)
I (17988) wifi:state: auth -> assoc (0)
I (17998) wifi:state: assoc -> run (10)
I (24098) wifi:state: run -> init (fc0)
I (24098) `wifi:new:<2,0>, old:<2,0>, ap:<255,255>, sta:<2,0>, prof:1`
I (24098) wifi station: retry to connect to the AP
I (24098) wifi station: connect to the AP fail
I (26508) wifi station: connect to the AP fail

**Reply by davefes** [reply]
I always put a 1uF from EN to ground.

Also, I don't know what your wifi connect script looks like.  I would suggest trying this example:
`https://github.com/glenn20/micropython-espnow-utils/tree/main/src/wifi.py`
Correction ... probably only works with his ESPNow images.

I haven't used idf-monitor so maybe watch things bootup from rshell.

**Comment by usafflyer1** [comment]
I registered just to say I was able to fix by selecting DIO. Default was QIO.

**Comment by waanito** [comment]
I may be way off course here but I am able to get mine (v3 not v1) to work without any mods. 
I have an esp32-wroom-32  on an esp32 devkitv1 board  but the Chip is ESP32-D0WD-V3 (revision v3.1) not OP's V1 
on macos  terminal:

esptool.py erase_flash
...
Chip erase completed successfully in 2.9s
Hard resetting via RTS pin...

esptool.py write_flash 0x1000 esp32-20230426-v1.20.0.bin
esptool.py v4.6.2
Found 4 serial ports
Serial port /dev/cu.usbserial-0001
/dev/cu.usbserial-0001 failed to connect: Could not open /dev/cu.usbserial-0001, the port doesn't exist
Serial port /dev/cu.SLAB_USBtoUART
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting.......
Detecting chip type... ESP32
Chip is ESP32-D0WD-V3 (revision v3.1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: b0:a7:32:81:df:50
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00001000 to 0x0017ffff...
Compressed 1566528 bytes to 1034676...
Wrote 1566528 bytes (1034676 compressed) at 0x00001000 in 91.3 seconds (effective 137.2 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

then:
esptool.py verify_flash 0x1000 esp32-20230426-v1.20.0.bin 
...
Verifying 0x17e740 (1566528) bytes @ 0x00001000 in flash against esp32-20230426-v1.20.0.bin...
-- verify OK (digest matched)
Hard resetting via RTS pin...

then using ESP MPY-JAMA ide 
MicroPython >>> import network
MicroPython >>> ap = network.WLAN(network.AP_IF)

no errors

Hope this helps someone
